src/python/pants/engine/process.py

```diff
@@ -50,6 +51,57 @@ class ProcessCacheScope(Enum):
     PER_SESSION = "per_session"
 
 
+@dataclass(frozen=True)
+class ProcessConcurrency:
+    kind: Literal["exactly", "range", "exclusive"]
+    min: int | None = None
+    max: int | None = None
+
+    def __post_init__(self):
+        if self.min is not None and self.min < 1:
+            raise ValueError(f"min concurrency must be >= 1, got {self.min}")
+        if self.max is not None and self.max < 1:
+            raise ValueError(f"max concurrency must be >= 1, got {self.max}")
+        if self.min is not None and self.max is not None and self.min > self.max:
+            raise ValueError(
+                f"min concurrency must be <= max concurrency, got {self.min} and {self.max}"
+            )
+        if self.kind == "exactly" and self.min != self.max:
+            raise ValueError(
+                f"exactly concurrency must have min and max equal, got {self.min} and {self.max}"
+            )
+
+    @staticmethod
+    def range(max: int, min: int = 1):
```

Comment:
Minor tweaks on the python side since last time @benjyw The ordering of the arguments has been reversed for `range`, this is because I wanted range to be 1-for-1 with the old `concurrency_available`. So the lower bound defaults to 1, and you can specify the upper bound unnamed `ProcessConcurrency.range(5)`

Iff you use `range` you must also use the template var `{pants_concurrency}` in the `argv`, just like `concurrency_available`

Reply:
do you think there's anything we could change to clarify it? Or leave that for a follow up after users are exposed to it?

Reply:
one idea: make it `range(min: int, max: int)` (no optional args) and add another method `.at_most(max, min=1)` which is the shorthand with reversed arguments

Reply:
Yeah, let's leave it for now